rust/pattern/constraint/type_constraint/owns_constraint.rs

```diff
@@ -0,0 +1,152 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::{
+    common::token::Constraint::{IsKey, Owns},
+    Label, Type, TypeVariable, TypeVariableBuilder, UnboundVariable,
+};
+use std::fmt;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum IsKeyAttribute {
+    Yes,
+    No,
+}
+
+impl From<bool> for IsKeyAttribute {
+    fn from(is_key: bool) -> Self {
+        match is_key {
+            true => IsKeyAttribute::Yes,
+            false => IsKeyAttribute::No,
+        }
+    }
+}
+
+pub const KEY: IsKeyAttribute = IsKeyAttribute::Yes;
+
+impl fmt::Display for IsKeyAttribute {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if *self == IsKeyAttribute::Yes {
+            write!(f, " {}", IsKey)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct OwnsConstraint {
+    pub attribute_type: TypeVariable,
+    pub overridden_attribute_type: Option<TypeVariable>,
+    pub is_key: IsKeyAttribute,
+}
+
+impl OwnsConstraint {
+    fn new(
+        attribute_type: TypeVariable,
+        overridden_attribute_type: Option<TypeVariable>,
+        is_key: IsKeyAttribute,
+    ) -> Self {
+        OwnsConstraint { attribute_type, overridden_attribute_type, is_key }
+    }
+}
+
+impl From<&str> for OwnsConstraint {
+    fn from(attribute_type: &str) -> Self {
+        OwnsConstraint::from((attribute_type, IsKeyAttribute::No))
+    }
+}
+
+impl From<(&str, IsKeyAttribute)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (&str, IsKeyAttribute)) -> Self {
+        OwnsConstraint::from((Label::from(attribute_type), is_key))
+    }
+}
+
+impl From<(String, IsKeyAttribute)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (String, IsKeyAttribute)) -> Self {
+        OwnsConstraint::from((Label::from(attribute_type), is_key))
+    }
+}
+
+impl From<(Label, IsKeyAttribute)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (Label, IsKeyAttribute)) -> Self {
+        OwnsConstraint::from((UnboundVariable::hidden().type_(attribute_type).unwrap(), is_key))
+    }
+}
+
+impl From<(UnboundVariable, IsKeyAttribute)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (UnboundVariable, IsKeyAttribute)) -> Self {
+        OwnsConstraint::from((attribute_type.into_type(), is_key))
+    }
+}
+
+impl From<TypeVariable> for OwnsConstraint {
+    fn from(attribute_type: TypeVariable) -> Self {
+        OwnsConstraint::new(attribute_type, None, IsKeyAttribute::No)
+    }
+}
+
+impl From<(TypeVariable, IsKeyAttribute)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (TypeVariable, IsKeyAttribute)) -> Self {
+        OwnsConstraint::new(attribute_type, None, is_key)
+    }
+}
+
+impl From<Type> for OwnsConstraint {
+    fn from(role_type: Type) -> Self {
+        OwnsConstraint::from(role_type.into_type_variable())
+    }
+}
```

Comment:
Many type constraints were extended to now accept `Type` as well (`Either<Pair<String, String>, UnboundVariable>` in Java parlance).